### ISSUE TITLE
np.nanmean executes eagerly on cubed arrays but lazily on dask arrays

### ISSUE THREAD

**Issue by TomNicholas** [cubed-dev/cubed#153]
See image for demonstration.

![Screenshot from 2023-03-14 19-26-13](https://user-images.githubusercontent.com/35968931/225164219-0125df14-e3f9-46ee-85c2-8ec523093ec1.png)

`np.nanmean` is called by xarray's `.mean()` method when `skipna=True`, which is the default.

**Comment by tomwhite** [comment]
Dask supports the `__array_ufunc__` and `__array_function__` protocols which is why it works lazily in this case. Cubed doesn't, and `nanmean` is not in the Array API, so `np.nanmean` will coerce the Cubed array to a NumPy array eagerly.

There's a bit of a discussion about the nanfunctions in https://github.com/data-apis/array-api/issues/170, including  workarounds, but I'm not sure if that would help you @TomNicholas?

**Comment by TomNicholas** [comment]
> nanmean is not in the Array API

Yeah I noticed that - what cubed does here is totally reasonable. I wasn't sure whether to raise this on cubed or xarray.

> There's a bit of a discussion about the nanfunctions in https://github.com/data-apis/array-api/issues/170,

Thanks! [This comment](https://github.com/data-apis/array-api/issues/170#issuecomment-827548927) mentions using the `where` kwarg to mean, which I presume is something we could change xarray to do instead of calling `nanmean` (though I expect `nanmean` is probably more performant?). The `where` kwarg also isn't in the Array API standard either though by the looks of it. They say there is another discussion about nan-aware functions but I can't find it by searching that repo :confused:

**Comment by dcherian** [comment]
see https://github.com/pydata/xarray/pull/7424
